middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,101 @@
+
+    received_data = b''
+
+    while True:
+        next_bytes = connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        for next_byte in next_bytes:
+            if next_bytes == FRAME_SEPARATOR:
+                yield received_data
+                received_data = b''
+            else:
+                received_data += bytes([next_byte])
+
+
+def unescape_stream(connection: socket.socket) -> Iterator[Optional[bytes]]:
+    """
+    Middle level receiver which determines if received data has broken escaping and yields None to indicate it,
+    or reverses escaping and yields unescaped frame.
+    """
+    assert isinstance(connection, socket.socket)
+
+    for unescaped_data in split_stream(connection=connection):
+        assert isinstance(unescaped_data, bytes)
+
+        # If there is an escape character which is not a valid escape sequence,
+        # yield special value indicating a frame with broken escaping and continue the outer loop.
+        # This can be checked by comparing sum of occurrences of escape character
+        # to sum of occurrences of escape sequences. If they are equal, data is escaped correctly.
+        if (
+            unescaped_data.count(ESCAPE_CHARACTER) !=
+            sum([unescaped_data.count(escape_sequence) for escape_sequence in ESCAPE_SEQUENCES.values()])
+        ):
+            yield None
+            continue
+
+        yield escape_decode_raw_message(unescaped_data)
+
+
+def receive_frame(connection: socket.socket, public_key: bytes) -> Iterator[Optional[AbstractMiddlemanMessage]]:
+    """
+    Top level receiver which determines if received data indicates an invalid escape sequence and yields it,
+    or if received data is valid Middleman protocol message and yields instance of it.
+    """
+    assert isinstance(connection, socket.socket)
+    assert isinstance(public_key, bytes)
+
+    for received_frame in unescape_stream(connection=connection):
+        assert isinstance(received_frame, (bytes, type(None)))
+
+        # If you have just received the value indicating an invalid escape sequence,
+        # yield a special value indicating an invalid escape sequence and continue.
+        if received_frame is None:
+            yield received_frame
+
+        # Otherwise create an instance of the frame class and yield it.
+        else:
+            yield AbstractMiddlemanMessage.deserialize(
+                raw_message=received_frame,
+                public_key=public_key,
+            )
+
+
+def prepare_stream(raw_message: AbstractMiddlemanMessage, private_key: bytes) -> bytes:
+    """
+    Helper for preparing AbstractMiddlemanMessage for sending through socket stream by
+    * serializing,
+    * escaping,
+    * adding frame separator.
+    """
+    assert isinstance(raw_message, AbstractMiddlemanMessage)
+    assert isinstance(private_key, bytes)
+
+    return append_frame_separator(
+        escape_encode_raw_message(
+            raw_message.serialize(
```

Comment:
I think this should also send the message over the connection - because  it's meant to be the reverse of the generators above. Or we should have another helper that does it. It should all be symmetric.

Reply:
:+1: done